diff -Naur linux-3.3.2-org/arch/arm/mach-omap2/omap_hwmod_44xx_data.c linux-3.3.2-ths/arch/arm/mach-omap2/omap_hwmod_44xx_data.c
--- linux-3.3.2-org/arch/arm/mach-omap2/omap_hwmod_44xx_data.c	2012-04-15 10:38:04.000000000 +0200
+++ linux-3.3.2-ths/arch/arm/mach-omap2/omap_hwmod_44xx_data.c	2012-04-16 12:08:45.000000000 +0200
@@ -2262,7 +2262,8 @@
 };
 
 static struct omap_i2c_dev_attr i2c_dev_attr = {
-	.flags	= OMAP_I2C_FLAG_BUS_SHIFT_NONE,
+	.flags	= OMAP_I2C_FLAG_BUS_SHIFT_NONE |
+			OMAP_I2C_FLAG_RESET_REGS_POSTIDLE,
 };
 
 /* i2c1 */
