--- arch/arm/mach-omap2/devices.c	2012-06-19 21:47:13.000000000 +0530
+++ arch/arm/mach-omap2/devices.c	2012-06-22 18:57:11.894677483 +0530
@@ -682,6 +682,7 @@
 	omap_init_audio();
 	omap_init_mcpdm();
 	omap_init_camera();
+	omap3_init_camera(NULL);
 	omap_init_mbox();
 	omap_init_mcspi();
 	omap_init_pmu();
--- arch/arm/mach-omap2/board-overo.c	2012-06-19 21:47:13.000000000 +0530
+++ arch/arm/mach-omap2/board-overo.c	2012-07-06 13:20:43.386634921 +0530
@@ -513,6 +513,22 @@
 	.vmmc1		= &overo_vmmc1,
 };
 
+#define LM3553_SLAVE_ADDRESS	0x53
+#define OV3640_I2C_ADDR		(0x78 >> 1)
+int omap3evm_ov3640_platform_data;
+int lm3553_platform_data;
+
+static struct i2c_board_info __initdata omap3_i2c_boardinfo_3[] = {
+	{
+		I2C_BOARD_INFO("ov3640", OV3640_I2C_ADDR),
+		.platform_data = &omap3evm_ov3640_platform_data,
+	},
+	{
+		I2C_BOARD_INFO("lm3553",LM3553_SLAVE_ADDRESS),
+		.platform_data = &lm3553_platform_data,
+	},
+};
+
 static int __init overo_i2c_init(void)
 {
 	u32 pdata_flags = 0;
@@ -542,7 +558,7 @@
 
 	omap3_pmic_init("tps65950", &overo_twldata);
 	/* i2c2 pins are used for gpio */
-	omap_register_i2c_bus(3, 400, NULL, 0);
+	omap_register_i2c_bus(3, 100, omap3_i2c_boardinfo_3, ARRAY_SIZE(omap3_i2c_boardinfo_3));
 	return 0;
 }
 
--- drivers/media/video/omap/omap_vout.c	2012-06-19 21:47:28.000000000 +0530
+++ drivers/media/video/omap/omap_vout.c	2012-07-03 18:11:44.616553590 +0530
@@ -2232,7 +2232,7 @@
 	if (ret)
 		goto probe_err2;
 
-	for (i = 0; i < vid_dev->num_displays; i++) {
+	for (i = 0; i < vid_dev->num_displays-1; i++) {
 		struct omap_dss_device *display = vid_dev->displays[i];
 
 		if (display->driver->update)

